File: 2023/go/pkg/day3/problem1.go
package day3

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

func SolveProblem1(lines []string) string {
	sum := 0

	for i, line := range lines {
		var temp strings.Builder

		for j, r := range line {
			if unicode.IsDigit(r) {
				temp.WriteRune(r)
			}

			if (!unicode.IsDigit(r) || j == len(line)-1) && temp.Len() > 0 {
				number, err := strconv.Atoi(temp.String())

				if err != nil {
					panic(err)
				}

				if isAdjacentToSymbol(j-temp.Len(), i, temp.String(), lines) {
					sum += number
				}

				temp.Reset()
			}
		}
	}

	return fmt.Sprint(sum)
}

func isAdjacentToSymbol(x, y int, num string, lines []string) bool {
	for dy := -1; dy <= 1; dy++ {
		for dx := -1; dx <= len(num); dx++ {
			ny, nx := y+dy, x+dx
			if ny >= 0 && ny < len(lines) && nx >= 0 && nx < len(lines[ny]) && !unicode.IsDigit(rune(lines[ny][nx])) && lines[ny][nx] != '.' {
				return true
			}
		}
	}
	return false
}

File: 2023/rust/src/day3/problem1.rs
pub fn solve(lines: Vec<&str>) -> String {
    let mut sum = 0;

    for (i, line) in lines.iter().enumerate() {
        let mut temp = String::new();

        for (j, char) in line.chars().enumerate() {
            if char.is_digit(10) {
                temp.push(char);
            }

            if (!char.is_digit(10) || j == line.len() - 1) && temp.len() > 0 {
                if is_adjecent_to_symbol(&temp, j - temp.len(), i, &lines) {
                    let num = temp.parse::<i32>().expect("Failed to parse number :(");
                    sum += num;
                }

                temp.clear();
            }
        }
    }

    return sum.to_string();
}

fn is_adjecent_to_symbol(num: &String, x: usize, y: usize, lines: &Vec<&str>) -> bool {
    for dy in -1..2 {
        for dx in -1..num.len() as i32 + 1 {
            let x = x as i32 + dx;
            let y = y as i32 + dy;

            if y < 0 || x < 0 {
                continue;
            }

            let x = x as usize;
            let y = y as usize;

            if y >= lines.len() || x >= lines[y].len() {
                continue;
            }

            let char = lines[y].chars().nth(x).unwrap();

            if !char.is_digit(10) && char != '.' {
                return true;
            }
        }
    }

    false
}
